src/providers/hana/qgshanaconnection.cpp

```diff
@@ -0,0 +1,621 @@
+                    .arg( layerProperty.schemaName, layerProperty.tableName, layerProperty.geometryColName );
+      ResultSetRef rsSrid = stmt->executeQuery( reinterpret_cast<const char16_t *>( sql.unicode() ) );
+      ret = rsSrid->next() ? *rsSrid->getInt( 1 ) : -1;
+      rsSrid->close();
+    }
+    else
+    {
+      QString sql = QStringLiteral( "SELECT %1.ST_SRID() FROM %2.%3 WHERE %1 IS NOT NULL LIMIT %4" )
+                    .arg( QgsHanaUtils::quotedIdentifier( layerProperty.geometryColName ),
+                          QgsHanaUtils::quotedIdentifier( layerProperty.schemaName ),
+                          QgsHanaUtils::quotedIdentifier( layerProperty.tableName ),
+                          QString::number( GEOM_TYPE_SELECT_LIMIT ) );
+      ResultSetRef rsSrid = stmt->executeQuery( reinterpret_cast<const char16_t *>( sql.unicode() ) );
+      int srid = -1, prevSrid = -1;
+      while ( rsSrid->next() )
+      {
+        srid = *rsSrid->getInt( 1 );
+        if ( prevSrid != -1 && srid != prevSrid )
+        {
+          srid = -1;
+          break;
+        }
+        prevSrid = srid;
+      }
+
+      ret = srid;
+      rsSrid->close();
+    }
+  }
+  catch ( const Exception &ex )
+  {
+    throw QgsHanaException( ex.what() );
+  }
+
+  return ret;
+}
+
+QStringList QgsHanaConnection::getLayerPrimaryeKeys( const QgsHanaLayerProperty &layerProperty )
+{
+  QStringList ret;
+
+  try
+  {
+    DatabaseMetaDataRef dbmd = mConnection->getDatabaseMetaData();
+    ResultSetRef rsPrimaryKeys = dbmd->getPrimaryKeys( nullptr,
+                                 layerProperty.schemaName.toStdString().c_str(),
+                                 layerProperty.tableName.toStdString().c_str() );
+
+    while ( rsPrimaryKeys->next() )
+    {
+      auto keyName = rsPrimaryKeys->getNString( 4 );
+      if ( !keyName.isNull() )
+      {
+        QString clmName = QgsHanaUtils::toQString( keyName );
+        ResultSetRef rsColumns = dbmd->getColumns( nullptr,
+                                 layerProperty.schemaName.toStdString().c_str(),
+                                 layerProperty.tableName.toStdString().c_str(),
+                                 clmName.toStdString().c_str() );
+
+        if ( rsColumns->next() )
+        {
+          Short dataType = rsColumns->getShort( 5 );
+          if ( !dataType.isNull() )
+          {
+            short dt = *dataType;
+            if ( dt == SQLDataTypes::TinyInt || dt == SQLDataTypes::SmallInt ||
+                 dt == SQLDataTypes::Integer || dt == SQLDataTypes::BigInt )
+              ret << clmName;
+          }
+        }
+        rsColumns->close();
+      }
+    }
+    rsPrimaryKeys->close();
+  }
+  catch ( const Exception &ex )
+  {
+    throw QgsHanaException( ex.what() );
+  }
+
+  return ret;
+}
+
+QgsWkbTypes::Type QgsHanaConnection::getLayerGeometryType( const QgsHanaLayerProperty &layerProperty )
```

Comment:
Do we also need to consider different types for geometries with z- and m-coordinates?

Reply:
Done in b6ffb17.